src/Files.Launcher/MessageHandlers/Win32MessageHandler.cs

```diff
@@ -234,6 +242,92 @@ await Win32API.SendMessageAsync(connection, new ValueSet()
+                    break;
+            }
+        }
+
+        private async Task ParseWatchDirectoryActionAsync(PipeStream connection, Dictionary<string, object> message, string action)
+        {
+            switch (action)
+            {
+                case "start":
+                    {
+                        var res = new ValueSet();
+                        var folderPath = (string)message["folderPath"];
+                        if (Directory.Exists(folderPath))
+                        {
+                            var watcher = new FileSystemWatcher
+                            {
+                                Path = folderPath,
+                                Filter = "*.*",
+                                NotifyFilter = NotifyFilters.LastWrite | NotifyFilters.FileName | NotifyFilters.DirectoryName
+                            };
+                            watcher.Created += DirectoryWatcher_Changed;
+                            watcher.Deleted += DirectoryWatcher_Changed;
+                            watcher.Renamed += DirectoryWatcher_Changed;
+                            watcher.EnableRaisingEvents = true;
+                            res.Add("watcherID", watcher.GetHashCode());
+                            dirWatchers.Add(watcher);
+                        }
+                        await Win32API.SendMessageAsync(connection, res, message.Get("RequestID", (string)null));
+                    }
+                    break;
+
+                case "cancel":
+                    {
+                        var watcherID = (long)message["watcherID"];
+                        var watcher = dirWatchers.SingleOrDefault(x => x.GetHashCode() == watcherID);
+                        if (watcher != null)
+                        {
+                            dirWatchers.Remove(watcher);
+                            watcher.Dispose();
+                        }
+                    }
+                    break;
+            }
+        }
+
+        private async void DirectoryWatcher_Changed(object sender, FileSystemEventArgs e)
+        {
+            System.Diagnostics.Debug.WriteLine($"Directory watcher event: {e.ChangeType}, {e.FullPath}");
+            if (connection?.IsConnected ?? false)
+            {
+                var response = new ValueSet()
+                {
+                    { "FileSystem", Path.GetDirectoryName(e.FullPath) },
+                    { "Name", e.Name },
+                    { "Path", e.FullPath },
+                    { "Type", e.ChangeType.ToString() },
+                    { "WatcherID", sender.GetHashCode() },
+                };
+                if (e.ChangeType == WatcherChangeTypes.Created)
+                {
+                    using var folderItem = SafetyExtensions.IgnoreExceptions(() => new ShellItem(e.FullPath));
+                    if (folderItem == null) return;
+                    var shellFileItem = ShellFolderExtensions.GetShellFileItem(folderItem);
```

Comment:
I can't test this since the file paste stops due to an exception. First I tried just installing the azure build, and when I saw the pasting didn't complete I tried debugging with VS, and this is the result:

![image](https://user-images.githubusercontent.com/11770760/161769587-15bc72eb-6553-4d2d-803b-e56d54d62407.png)

I assume that the issue will be on this line since it's happening on GetShellFileItem, maybe is necessary another checking before calling the method?

Reply:
@gave92

Reply:
Could you confirm the following?
- that the file is correctly pasted (it appears if you refresh)
- what kind of drive you're copying to (SMB share?) and from where and what kind of item
- if renaming has the same issue while deleting works ok

Reply:
@gave92 
> I'm having trouble reproducing. Could you confirm the following?
> 
> * if the file is correctly pasted (it appears if you refresh)

The .exe file not, have the size but corrupted:
![image](https://user-images.githubusercontent.com/11770760/161854202-dd634a27-7158-453d-a1bf-52aaf794eb92.png)

About image files, the first get copied correctly, but after trying pasting another Files will crash and it won't get copied.

> * what kind of drive you're copying to (another PC shared folder?)

Synology shared folder using SMB protocol.

> * what kind of item you're copying

In the original test I've tried with an .exe, .gif and .jpg. Now I also tried with a .zip file and this format seems to work as expected.
BTW, I just realize that the .jpg/.gif error and .exe error is not the same. JPG/GIF throw this one:
![image](https://user-images.githubusercontent.com/11770760/161853728-d504180c-664d-407d-9b01-6a50dc90f3c6.png)
('The process does not have access to the file because it is being used by another process. )

> * if renaming and deleting work ok

Yes, both operations seems to work ok and refresh as expected.

> 
> Also, could you check what happens if you change those lines (45-47) to:
> 
> ```cs
> //folderItem.Properties.TryGetValue<string>(
> //    Ole32.PROPERTYKEY.System.ItemNameDisplay, out var fileName);
> var fileName = Path.GetFileName(folderItem.Name); // Original file name
> ```

In this case, the error will move to lines 61-62. For continuing the test I modified the code to continue without crashing with this:
![image](https://user-images.githubusercontent.com/11770760/161862297-70f914c7-6ffd-4631-ae39-0d0bc19ec4c6.png)

With all this changes, the setup seems to work:
![image](https://user-images.githubusercontent.com/11770760/161862620-5f45e145-f6eb-43c4-8e8d-3f4dc5ceebea.png)

Also it got automatically refresh and pasting 2 images one after the other won't make Files crash. The only issue is the expected, no file data:
![image](https://user-images.githubusercontent.com/11770760/161862856-ecad165e-5dcd-456d-8a15-9359000d3dbc.png)

After a manual refresh:
![image](https://user-images.githubusercontent.com/11770760/161862917-d43de992-b512-4c30-9713-582b8909fc43.png)

Reply:
Thanks I think the issue is that the created event is raised before the file has finished writing.

Reply:
7b24ce7 should fix this issue.